File: problems/FastFourierTransform/FFT.cpp
#include <iostream>
#include <vector>
#include <complex>
#include <cmath>
#include <omp.h>
#include <iomanip>
#include <chrono>

const double PI = acos(-1);
using Complex = std::complex<double>;
using namespace std;

inline Complex omega(int k, int n) {
    double angle = -2.0 * PI * k / n;
    return Complex(cos(angle), sin(angle));
}

void fft(vector<Complex>& a, vector<Complex>& out, int start, int n, int step) {
    if (n == 1) {
        out[start / step] = a[start];
        return;
    }

    int half = n / 2;
    vector<Complex> even(half), odd(half);

    if (n > 2048) {
        #pragma omp task shared(a)
        fft(a, even, start, half, 2 * step);

        #pragma omp task shared(a)
        fft(a, odd, start + step, half, 2 * step);

        #pragma omp taskwait
    } 
    
    else 
    {
        fft(a, even, start, half, 2 * step);
        fft(a, odd, start + step, half, 2 * step);
    }

    for (int k = 0; k < half; ++k) {
        Complex w = omega(k, n) * odd[k];
        out[k] = even[k] + w;
        out[k + half] = even[k] - w;
    }
}

int main() {
    int n = 1 << 20; // 2^20 = 1,048,576 points
    vector<Complex> input(n, 1); // All values set to 1
    vector<Complex> output(n);

    omp_set_num_threads(4); // Set number of threads

    auto start_time = chrono::high_resolution_clock::now();

    #pragma omp parallel
    {
        #pragma omp single
        {
            fft(input, output, 0, n, 1);
        }
    }

    auto end_time = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end_time - start_time;

    cout << "Parallel FFT done in: " << duration.count() << " seconds" << endl;

    return 0;
}

File: problems/FastFourierTransform/FFT.ml
module Task = Domainslib.Task

let pi = 4.0 *. atan 1.0

type complex = float * float

let complex_add (a, b) (c, d) = (a +. c, b +. d)
let complex_sub (a, b) (c, d) = (a -. c, b -. d)
let complex_mul (a, b) (c, d) = (a *. c -. b *. d, a *. d +. b *. c)

let omega k n =
  let angle = -.2.0 *. pi *. float_of_int k /. float_of_int n in
  (cos angle, sin angle)

  let rec fft pool a out start n step =
    if n = 1 then
      Array.set out (start / step) a.(start)
    else
      let half = n / 2 in
      let even = Array.make half (0.0, 0.0) in
      let odd  = Array.make half (0.0, 0.0) in
  
      if n > 2048 then begin
        let t1 = Task.async pool (fun () -> fft pool a even start half (2 * step)) in
        let t2 = Task.async pool (fun () -> fft pool a odd (start + step) half (2 * step)) in
        Task.await pool t1;
        Task.await pool t2
      end else begin
        fft pool a even start half (2 * step);
        fft pool a odd (start + step) half (2 * step)
      end;
  
      for k = 0 to half - 1 do
        let w = complex_mul (omega k n) odd.(k) in
        out.(k) <- complex_add even.(k) w;
        out.(k + half) <- complex_sub even.(k) w
      done  

let () =
  let n = 1 lsl 20 in
  let input = Array.make n (1.0, 0.0) in
  let output = Array.make n (0.0, 0.0) in

  let pool = Task.setup_pool ~num_domains:4 () in
  let t0 = Unix.gettimeofday () in

  (* fft pool input output 0 n 1; *)
  let res = Task.run pool (fun _ -> fft pool input output 0 n 1) in

  let t1 = Unix.gettimeofday () in
  Task.teardown_pool pool;

  Printf.printf "Parallel FFT done in: %.6f seconds\n%!" (t1 -. t0)
